### DISCUSSION TITLE
Renovate bot self-hosted with Docker (podman) example seems broken

### DISCUSSION THREAD

**Discussion by yannickincyberatlantis** [renovatebot/renovate#28118] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitea

### Please tell us more about your question or problem

When running the example from here (using podman rather than docker!): https://docs.renovatebot.com/examples/self-hosting/#docker

With this `config.js` (I don't believe it matters though):
```js
module.exports = {
  platform: 'gitea',
  endpoint: 'https://gitea.ocram85.com',
  gitAuthor: 'renovate-bot <renovate@ocram85.com>',
  autodiscover: true,
  prHourlyLimit: 10,
  automerge: false,
  major: {
    automerge: false,
  },
  automergeType: 'pr',
  automergeStrategy: 'squash',
  labels: ['renovate'],
  dependencyDashboardLabels: ['renovate'],
  githubTokenWarn: false,
  semanticCommits: 'enabled',
  autodiscoverFilter: ['OCram85/*', 'CodeServer/*', 'renovate-bot/*'],
}
```

(Taken from: https://gitea.ocram85.com/renovate-bot/config)

I get the following error:
```bash
/tmp/renovate via  v20.10.0 took 2s 
❯ podman run --rm -v config.js:/usr/src/app/config.js ghcr.io/renovatebot/renovate
FATAL: You must configure a GitHub token
 INFO: Renovate was run at log level "info". Set LOG_LEVEL=debug in environment variables to see extended debug logs.
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "M58D0eDvPNoQ7JcGUfgBW",
           "msg": "You must configure a GitHub token"
         }
       ]
```

When enabling debug logging it looks like Renovate doesn't pick up on the config file I supplied (**DEBUG: No config file found on disk - skipping**):

```bash
/tmp/renovate via  v20.10.0 took 2s 
❯ podman run --rm -v config.js:/usr/src/app/config.js -e LOG_LEVEL=debug ghcr.io/renovatebot/renovate
DEBUG: Using RE2 regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in config.js
DEBUG: No config file found on disk - skipping
DEBUG: File config
       "config": {}
DEBUG: CLI config
       "config": {}
DEBUG: Env config
       "config": {"hostRules": []}
DEBUG: Combined config
       "config": {"hostRules": []}
DEBUG: Enabling forkProcessing while in non-autodiscover mode
DEBUG: Found valid git version: 2.43.2
DEBUG: Setting global hostRules
FATAL: You must configure a GitHub token
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "7ls7Y3Cstfpr7Lym4UrIk",
           "msg": "You must configure a GitHub token"
         }
       ]
```

The file config parser (`file.ts`) fairly recently had some updates, perhaps those broke it?

Anyway, if I supply `RENOVATE_CONFIG_FILE="config.js"` (note that this should be the same value as the default from `file.ts`) I get a different error:

```bash
/tmp/renovate via  v20.10.0 
❯ podman run --rm -v config.js:/usr/src/app/config.js -e RENOVATE_CONFIG_FILE="config.js" ghcr.io/renovatebot/renovate
FATAL: Error parsing config file
```

To me this would indicate the `config.js` is incorrect, however when I run the config validator it doesn't return any errors:
```bash
/tmp/renovate via  v20.10.0 took 2s 
❯ npx --yes --package renovate -- renovate-config-validator config.js
 INFO: Validating config.js
 INFO: Config validated successfully
```

I'm lost as to what the exact issue is here, has anyone ran into this as well and have any suggestions on what I'm doing wrong? Besides that it would be nice to have some more detailed information (perhaps only when running with `debug` or `trace`) on *what* kind of parsing error is occurring.

**Edit**:

When running Renovate locally on my machine using node.js it seems to pass the config parsing stage:

```bash
/tmp/renovate via  v20.10.0 
❯ LOG_LEVEL=debug RENOVATE_CONFIG_FILE="config.js" ./node_modules/.bin/renovate
DEBUG: Using RE2 regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in config.js
DEBUG: File config
       "config": {
         "platform": "gitea",
         "endpoint": "https://gitea.ocram85.com",
         "gitAuthor": "renovate-bot <renovate@ocram85.com>",
         "autodiscover": true,
         "prHourlyLimit": 10,
         "automerge": false,
         "major": {"automerge": false},
         "automergeType": "pr",
         "automergeStrategy": "squash",
         "labels": ["renovate"],
         "dependencyDashboardLabels": ["renovate"],
         "githubTokenWarn": false,
         "semanticCommits": "enabled",
         "autodiscoverFilter": ["OCram85/*", "CodeServer/*", "renovate-bot/*"]
       }
DEBUG: CLI config
       "config": {}
DEBUG: Env config
       "config": {"hostRules": []}
DEBUG: Combined config
       "config": {
         "platform": "gitea",
         "endpoint": "https://gitea.ocram85.com",
         "gitAuthor": "renovate-bot <renovate@ocram85.com>",
         "autodiscover": true,
         "prHourlyLimit": 10,
         "automerge": false,
         "major": {"automerge": false},
         "automergeType": "pr",
         "automergeStrategy": "squash",
         "labels": ["renovate"],
         "dependencyDashboardLabels": ["renovate"],
         "githubTokenWarn": false,
         "semanticCommits": "enabled",
         "autodiscoverFilter": ["OCram85/*", "CodeServer/*", "renovate-bot/*"],
         "hostRules": []
       }
DEBUG: Adding trailing slash to endpoint
DEBUG: Found valid git version: 2.44.0
DEBUG: Setting global hostRules
FATAL: You must configure a Gitea personal access token
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "2dDH6_XxD7w2uwJ1oQIAQ",
           "msg": "You must configure a Gitea personal access token"
         }
       ]
```

The error above is expected, and is what I would expect to happen when running it via the container.

**Comment by rarkins** [comment]
docker run and podman run do not recognize "relative" filenames like "config.js". You need the full path name. Compare these two results:

```
$ docker run --rm -v config.js:/usr/src/app/config.js ghcr.io/renovatebot/renovate /bin/bash -c "cat config.js"
cat: config.js: Is a directory

$ docker run --rm -v "$(pwd)"/config.js:/usr/src/app/config.js ghcr.io/renovatebot/renovate /bin/bash -c "cat config.js"

module.exports = {
  // content
}
```

**Reply by yannickincyberatlantis** [reply, marked as answer]
Ah I see! I did try that as well, and that also didn't work. It looks like it's a combination of the issue you mentioned along with SELinux integration in Podman, I need to provide the `:z` flag to have Podman label the file properly. For more context see: https://docs.podman.io/en/latest/markdown/podman-run.1.html#volume-v-source-volume-host-dir-container-dir-options under "Labeling Volume Mounts".

TL;DR:

When using Podman along with SELinux you need to provide the `:z` flag _and_ provide the full path to `config.js` like @rarkins mentioned:

```bash
podman run --rm -v /path/to/config.js:/usr/src/app/config.js:z ghcr.io/renovatebot/renovate
```

Also note: If you run this command once, you can run it again without the `:z` flag, _until you change the `config.js` file again_, then you need the flag again.
